### ISSUE TITLE
False `invalid-argument-type` on `__getitem__` after `isinstance(x, np.ndarray)` narrows `Sequence[Any]` to `... & ndarray[object, object]`

### ISSUE THREAD

**Issue by adamtheturtle** [astral-sh/ty#3842]
## Summary

When `isinstance(x, np.ndarray)` narrows a value declared as `Sequence[Any]`, ty narrows it to the intersection `Sequence[Any] & ndarray[object, object]`. The synthesized `ndarray` partner has both type parameters filled with bare `object`, including the dtype slot (which expects `np.dtype[...]`). Indexing the narrowed value then fails overload resolution and produces a false `invalid-argument-type`. mypy accepts the same code.

## Reproduction

ty 0.0.51, numpy 2.2.6, Python 3.10:

```python
from collections.abc import Sequence
from typing import Any
import numpy as np

def f(x: Sequence[Any]) -> None:
    if isinstance(x, np.ndarray):
        reveal_type(x)          # Sequence[Any] & ndarray[object, object]
        mask = np.isin(x, [1])
        x[mask]                 # error: invalid-argument-type
```

ty reports two `invalid-argument-type` errors on `x[mask]`, one per member of the intersection:

- `Sequence[Any].__getitem__` cannot take an `ndarray[bool]` key (expected; `Sequence` has no mask overload)
- `ndarray[object, object].__getitem__` cannot take an `ndarray[bool]` key

The second is the false positive. mypy reports `Success: no issues found`.

## What looks wrong

The synthesized intersection partner is `ndarray[object, object]`: the dtype parameter is filled with bare `object` rather than a `dtype[...]`. A directly annotated object-dtype array does not have this problem:

```python
from numpy.typing import NDArray
def g(arr: NDArray[np.object_], mask: NDArray[np.bool_]) -> None:
    arr[mask]   # OK in ty
```

`NDArray[np.object_]` is `ndarray[Any, dtype[object_]]` (a proper `dtype[...]` parameter) and bool-mask indexing resolves fine. So the trigger is specifically the `object`-filled parameters produced by isinstance narrowing, not object dtype itself.

For comparison, narrowing a `Sequence[Any]` to a generic class whose type parameters are unbounded fills them with `Unknown` and wraps the result in `Top[...]`:

```python
from collections.abc import Sequence
from typing import Any, Generic, TypeVar, overload

T = TypeVar("T")
K = TypeVar("K")

class Arr(Sequence[Any], Generic[T, K]):
    @overload
    def __getitem__(self, i: int) -> Any: ...
    @overload
    def __getitem__(self, i: "Arr[Any, Any]") -> "Arr[T, K]": ...
    def __getitem__(self, i: Any) -> Any: ...
    def __len__(self) -> int: return 0

def f(x: Sequence[Any], mask: "Arr[Any, Any]") -> None:
    if isinstance(x, Arr):
        reveal_type(x)        # Sequence[Any] & Top[Arr[Unknown, Unknown]]
        x[mask]               # no error
```

Here the synthesized partner is `Top[Arr[Unknown, Unknown]]` (unknown parameters, wrapped in `Top[...]`) and `x[mask]` resolves without error. The numpy case (parameters with bounds/defaults) instead yields concrete `object` parameters and no `Top[...]` wrapper, which is what breaks method-overload resolution.

This is in the same family as #3817, but the manifestation differs: there it is the non-matching union member that becomes `X & Top[Generic[Unknown]]`; here it is the matching member that is narrowed to an intersection, and the synthesized partner uses `object` parameters that break `__getitem__` resolution.

## Environment

- ty 0.0.51
- numpy 2.2.6
- Python 3.10

**Comment by adamtheturtle** [comment]
Closed as an agent created this without asking me to, and I haven't investigated it.

**Comment by carljm** [comment]
This issue also looks valid, but it is a duplicate of #1109 -- the problem is that we top-materialize `dtype[Any]` to `dtype[object]`, even though it has an upper bound of `np.generic`.
